<a href="https://colab.research.google.com/github/rahiakela/mathematics-for-machine-learning/blob/main/deep-learning-book-maths/2_10_trace_of_matrix.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##The Trace of Matrix

This chapter is very light! I can assure you that you will read it in 1 minute! It is nice after the last two chapters that were quite big! We will see what is the Trace of a matrix. It will be needed for the last chapter on the Principal Component Analysis (PCA).

<img src="images/trace-matrix.png" width="200" alt="Calculating the trace of a matrix" title="Calculating the trace of a matrix">

<em>The trace of matrix</em>

##Setup

In [1]:
import numpy as np

from PIL import Image

import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Plot parameters
sns.set()
%pylab inline
pylab.rcParams['figure.figsize'] = (4, 4)
plt.rcParams['xtick.major.size'] = 0
plt.rcParams['ytick.major.size'] = 0
# Avoid inaccurate floating values (for inverse matrices in dot product for instance)
# See https://stackoverflow.com/questions/24537791/numpy-matrix-inversion-rounding-errors
np.set_printoptions(suppress=True)

Populating the interactive namespace from numpy and matplotlib


## Intuition

The trace is the sum of all values in the diagonal of a square matrix.

$$
{A}=
\begin{bmatrix}
    2 & 9 & 8 \\\\
    4 & 7 & 1 \\\\
    8 & 2 & 5
\end{bmatrix}
$$

$$
\mathrm{Tr}({A}) = 2 + 7 + 5 = 14
$$

Numpy provides the function `trace()` to calculate it:

In [3]:
A = np.array([
  [2, 9, 8],
  [4, 7, 1],
  [8, 2, 5]
])
A

array([[2, 9, 8],
       [4, 7, 1],
       [8, 2, 5]])

In [4]:
A_trace = np.trace(A)
A_trace

14

GoodFellow et al. explain that the trace can be used to specify the Frobenius norm of a matrix (see [2.5](https://hadrienj.github.io/posts/Deep-Learning-Book-Series-2.5-Norms/)). The Frobenius norm is the equivalent of the $L^2$ norm for matrices. It is defined by:

$$
||{{A}}||_F=\sqrt{\sum_{i,j}A^2_{i,j}}
$$

Take the square of all elements and sum them. Take the square root of the result. This norm can also be calculated with:

$$
||{{A}}||_F=\sqrt{Tr({{AA}^T})}
$$

We can check this. The first way to compute the norm can be done with the simple command `np.linalg.norm()`:

In [5]:
np.linalg.norm(A)

17.549928774784245

The Frobenius norm of ${A}$ is 17.549928774784245.

With the trace the result is identical:

In [6]:
np.sqrt(np.trace(A.dot(A.T)))

17.549928774784245

Since the transposition of a matrix doesn't change the diagonal, the trace of the matrix is equal to the trace of its transpose:

$$
Tr({A})=Tr({A}^T)
$$

In [7]:
assert np.trace(A) == np.trace(A.T)

##Trace of a product

$$
Tr({ABC}) = Tr({CAB}) = Tr({BCA})
$$

###Example 1: Trace property

Let's see an example of this property.

$$
{A}=
\begin{bmatrix}
    4 & 12 \\\\
    7 & 6
\end{bmatrix}
$$

$$
{B}=
\begin{bmatrix}
    1 & -3 \\\\
    4 & 3
\end{bmatrix}
$$

$$
{C}=
\begin{bmatrix}
    6 & 6 \\\\
    2 & 5
\end{bmatrix}
$$

In [8]:
A = np.array([
  [4, 12],
  [7, 6]
])

B = np.array([
  [1, -3], 
  [4, 3]
])

C = np.array([
  [6, 6], 
  [2, 5]
])

In [9]:
np.trace(A.dot(B).dot(C))

531

In [10]:
np.trace(C.dot(A).dot(B))

531

In [11]:
np.trace(B.dot(C).dot(A))

531

$$
{ABC}=
\begin{bmatrix}
    360 & 432 \\\\
    180 & 171
\end{bmatrix}
$$

$$
{CAB}=
\begin{bmatrix}
    498 & 126 \\\\
    259 & 33
\end{bmatrix}
$$

$$
{BCA}=
\begin{bmatrix}
    -63 & -54 \\\\
    393 & 594
\end{bmatrix}
$$

$$
Tr({ABC}) = Tr({CAB}) = Tr({BCA}) =  531
$$

In [12]:
assert np.trace(A.dot(B).dot(C)) == np.trace(C.dot(A).dot(B)) == np.trace(B.dot(C).dot(A)) == 531

## References

[Trace (linear algebra) - Wikipedia](https://en.wikipedia.org/wiki/Trace_(linear_algebra))

[Numpy Trace operator](https://docs.scipy.org/doc/numpy/reference/generated/numpy.trace.html)